src/plugins/expression_reveal_image/public/components/reveal_image_component.tsx

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { useRef, useState, useEffect, useCallback } from 'react';
+import { IInterpreterRenderHandlers } from 'src/plugins/expressions';
+import { NodeDimensions, RevealImageRendererConfig, Origin } from '../expression_renderers/types';
+import { isValidUrl } from '../../common/lib/url';
+import { elasticOutline } from '../../common/lib/elastic_outline';
+
+interface RevealImageComponentProps extends RevealImageRendererConfig {
+  handlers: IInterpreterRenderHandlers;
+  parentNode: HTMLElement;
+}
+
+interface ImageStyles {
+  width?: string;
+  height?: string;
+  clipPath?: string;
+}
+
+interface AlignerStyles {
+  backgroundImage?: string;
+}
+
+function RevealImageComponent({
+  handlers,
+  parentNode,
+  percent,
+  origin,
+  image,
+  emptyImage,
+}: RevealImageComponentProps) {
+  const [loaded, setLoaded] = useState(false);
+  const [dimensions, setDimensions] = useState<NodeDimensions>({
+    width: 1,
+    height: 1,
+  });
+
+  const imgRef = useRef<HTMLImageElement>(null);
+
+  // modify the top-level container class
+  parentNode.className = 'revealImage';
+
+  // set up the overlay image
+  const updateImageView = useCallback(() => {
+    if (imgRef.current) {
+      setDimensions({
+        height: imgRef.current.naturalHeight,
+        width: imgRef.current.naturalWidth,
+      });
+
+      setLoaded(true);
+      return handlers.done();
+    }
+  }, [imgRef, handlers]);
+
+  useEffect(() => {
+    handlers.event({ name: 'onResize', data: updateImageView });
```

Comment:
@crob611, what do you think about separating `emitters` and `listeners`? While I was discovering the code, related to handlers, I was very confused of treating `listeners` and `emitters` as events. Until I've read the implementation, I couldn't even imagine, that I can bind `listener` in that way.

Reply:
I've done this investigation. It is really relying on that event and there is no way of handing it in a different manner. The way of using this renderer is not allowing to listen to changes by props or in some other way. And I'm sure, there is no way of removing specific events in such an architecture. But, we can create `on(...)` method, which will put listeners if that events already exist in handlers and ignore, if not. I've implemented it in [current PR](https://github.com/elastic/kibana/pull/102714) (still in progress). I'll want to hear your opinion after finishing it.

Reply:
I would hold on the `on` stuff as I think our goal should be not having to do any listener registration.  

What this "onResize" really cares about is making sure the image is correctly adjusted as it's container is resized, but there are ways to go about that without needing to register this onResize function stuff. 

But, I'm good leaving the registration for now and we can always come back and clean it up.

Reply:
where is this event consumed and what is it used for ? if we just need to rerender when resized we usually used https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver

Reply:
@ppisljar, thank you, I'll try it.

Reply:
@ppisljar, your approach worked. I've used `useResizeObserver` from `@eui/elastic`. Thank you for the suggestion. I was afraid that observers are not supported in IE, but then I've seen that a lot of plugins are using it.